weavertest/multi.go

```diff
@@ -86,19 +90,81 @@ func initMultiProcess(ctx context.Context, t testing.TB, config string) weaver.I
+	//         --------        +----+        --------
+	//   fromWeaveletReader <--| OS |<-- fromWeaveletWriter
+	//     toWeaveletWriter -->|    |--> toWeaveletReader
+	//                         +----+
+	fromWeaveletReader, fromWeaveletWriter, err := os.Pipe()
 	if err != nil {
-		t.Fatal(err)
+		t.Fatal(fmt.Errorf("cannot create fromWeavelet pipe: %v", err))
+	}
+	toWeaveletReader, toWeaveletWriter, err := os.Pipe()
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create toWeavelet pipe: %v", err))
+	}
+
+	weaveletInfo := createWeaveletForMain(dep)
+
+	// Create a connection between the weavelet for the main process and the envelope.
+	econn, err := conn.NewEnvelopeConn(fromWeaveletReader, toWeaveletWriter, b, weaveletInfo)
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create envelope conn: %v", err))
 	}
+
+	// Create and run an envelope for the main process.
+	options := envelope.Options{
+		Restart:         envelope.Never,
+		Retry:           retry.DefaultOptions,
+		GetEnvelopeConn: func() *conn.EnvelopeConn { return econn },
+	}
+	e, err := envelope.NewEnvelope(weaveletInfo, dep.App, b, options)
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create envelope: %v", err))
+	}
+
+	go econn.Run()
+	go e.Run(ctx)
+
 	ctx = context.WithValue(ctx, runtime.BootstrapKey{}, runtime.Bootstrap{
-		ToWeaveletFd: toWeavelet,
-		ToEnvelopeFd: toEnvelope,
+		ToWeaveletFd: int(toWeaveletReader.Fd()),
+		ToEnvelopeFd: int(fromWeaveletWriter.Fd()),
 		TestConfig:   config,
 	})
 
 	t.Cleanup(func() {
 		cancelFunc()
+
+		// There is a very subtle bug we have to avoid. As explained in the
+		// official godoc [1], if an *os.File is garbage collected, the underlying
+		// file is closed. Thus, if toWeaveletReader or fromWeaveletWriter are
+		// garbage collected, the underlying pipes are closed.
+		//
+		// The main weavelet, which is running in a separate goroutine, is reading
+		// from and writing to these pipes, but it doesn't use toWeaveletReader or
+		// fromWeaveletWriter directly. Instead components to start., it constructs new files using the
+		// two pipe's file descriptors. This means that Go is free to garbage
+		// collect fromWeaveletWriter and toWeaveletWriter even though the
+		// underlying pipes are still being used by the main weavelet.
+		//
+		// If the pipes get closed while the main weavelet is running, then the
+		// weavelet can crash with all sorts of errors (e.g., bad pipe, resource
+		// temporarily unavailable, nil pointer dereferences, etc). You can run [2]
+		// locally to reproduce this behavior. If the program doesn't crash on your
+		// machine, try increasing the number of iterations.
+		//
+		// To avoid this, we have to ensure that the pipes are not garbage
+		// collected prematurely. This is done by holding a reference to the
+		// os.File in the closure passed to t.Cleanup
+		//
+		// [1]: https://pkg.go.dev/os#File.Fd
+		// [2]: https://go.dev/play/p/9JG7voL2oHP
+		_ = fromWeaveletWriter
+		_ = toWeaveletReader
+
```

Comment:
When `weavertest.Init` is called multple times by the same test package it may leak goroutines and keep resources around after the test has exited. I've found that can cause tests to flake, but I'm not sure if that will be a problem for serviceweaver applications.

I was thinking that maybe this would be a good place to close the `toWeaveletReader` and `fromWeavletReader`, so that the goroutines could shutdown gracefully. And then call `group.Wait` before returning so that all of the resources are stopped/cleaned up before the next test starts.

Do you think that will be important for testing serviceweaver applications?

---

I attempted that change, but it resulted in all kinds of problems (errors from the goroutines failing to `proto.Read`, and sometimes tests hanging until the timeout). I'll need to understand serviceweaver better before I'm able to fix those problems.

Reply:
> When weavertest.Init is called multple times by the same test package it may leak goroutines and keep resources around after the test has exited. I've found that can cause tests to flake, but I'm not sure if it will be a problem for serviceweaver applications.

Great point! I think it is a problem. As you've seen, the code right now does a bad job cleaning everything up when a test ends.  The `GetEnvelopeConn` option and the way it interacts with a babysitter is also confusing. I think cleaning all this code up would be amazing (but definitely tricky 😢). 

> I was thinking that maybe this would be a good place to close the toWeaveletReader and fromWeavletReader, so that the goroutines could shutdown gracefully. And then call errgroup.Group.Wait (or sync.WaitGroup.Wait`) before returning so that all of the resources are stopped/cleaned up before the next test starts. 

Woah, I've never seen `errgroup` before. Looks super useful!

> I attempted that change, but it resulted in all kinds of problems (errors from the goroutines failing to proto.Read, and sometimes tests hanging until the timeout). I'll need to understand serviceweaver better before I'm able to fix those problems.

I'm definitely not sure what's going wrong. This code is nasty and brittle. If I had to guess, I think it's related to the comment above about the "very subtle bug we have to avoid". I think the test ends, but a subprocess is still running with access to the `fromWeaveletWriter` and `toWeaveletReader` pipes. The subprocess never expects these to close; it closes them itself. So, if we do close them and the subprocess is running, it produces all kinds of nasty errors. I think leaving things as they are now sounds good to me.